modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/metadata/NodesMapping.java

```diff
@@ -0,0 +1,187 @@
+        this.flags = flags;
+    }
+
+    public List<UUID> nodes() {
+        return nodes;
+    }
+
+    public List<List<UUID>> assignments() {
+        return assignments;
+    }
+
+    public NodesMapping mergeWith(NodesMapping other) throws LocationMappingException {
+        byte flags = (byte) (this.flags | other.flags);
+
+        if ((flags & PARTIALLY_REPLICATED) == 0)
+            return new NodesMapping(U.firstNotNull(nodes, other.nodes), mergeAssignments(other, null), flags);
+
+        List<UUID> nodes;
+
+        if (this.nodes == null)
+            nodes = other.nodes;
+        else if (other.nodes == null)
+            nodes = this.nodes;
+        else
+            nodes = Commons.intersect(this.nodes, other.nodes);
+
+        if (nodes != null && nodes.isEmpty())
+            throw new LocationMappingException("Failed to map fragment to location.");
+
+        return new NodesMapping(nodes, mergeAssignments(other, nodes), flags);
+    }
+
+    public NodesMapping deduplicate() throws LocationMappingException {
+        if (assignments == null || !excessive())
+            return this;
+
+        HashSet<UUID> nodes0 = new HashSet<>();
+        List<List<UUID>> assignments0 = new ArrayList<>(assignments.size());
+
+        for (List<UUID> partNodes : assignments) {
+            UUID node = F.first(partNodes);
+
+            if (node == null)
+                throw new LocationMappingException("Failed to map fragment to location.");
+
+            assignments0.add(Collections.singletonList(node));
+            nodes0.add(node);
+        }
+
+        return new NodesMapping(new ArrayList<>(nodes0), assignments0, (byte)(flags | DEDUPLICATED));
+    }
+
+    public int[] partitions(UUID node) {
+        if (assignments == null)
+            return null;
+
+        GridIntList parts = new GridIntList(assignments.size());
+
+        for (int i = 0; i < assignments.size(); i++) {
+            List<UUID> assignment = assignments.get(i);
+            if (Objects.equals(node, F.first(assignment)))
+                parts.add(i);
+        }
+
+        return parts.array();
+    }
+
+    public boolean excessive() {
+        return (flags & DEDUPLICATED) == 0;
+    }
+
+    public boolean hasMovingPartitions() {
+        return (flags & HAS_MOVING_PARTITIONS) == HAS_MOVING_PARTITIONS;
+    }
+
+    public boolean hasReplicatedCaches() {
+        return (flags & HAS_REPLICATED_CACHES) == HAS_REPLICATED_CACHES;
+    }
+
+    public boolean hasPartitionedCaches() {
+        return (flags & HAS_PARTITIONED_CACHES) == HAS_PARTITIONED_CACHES;
+    }
+
+    public boolean partiallyReplicated() {
+        return (flags & PARTIALLY_REPLICATED) == PARTIALLY_REPLICATED;
+    }
+
+    private List<List<UUID>> mergeAssignments(NodesMapping other, List<UUID> nodes) throws LocationMappingException {
+        byte flags = (byte) (this.flags | other.flags); List<List<UUID>> left = assignments, right = other.assignments;
```

Comment:
Let's split this into 2 lines at least.
Declaring several variables of different type in one-liner is completely non-readable.